commit 38bbf023b6178888ec7f1750185e2f8d5a29176a
Author: Andy Whitcroft <apw@canonical.com>
Date:   Thu Jul 26 11:25:52 2012 +0100

    UBUNTU: SAUCE: (no-up) rds_ib_send() -- prevent local pings triggering BUG_ON()
    
    Pining localhost on an infiniband connection can trigger a BUG_ON()
    and cause a denial of service.
    
    Fix identified by comparison of the RHEL source rpms.
    
    CVE-2012-2372
    BugLink: http://bugs.launchpad.net/bugs/1016299
    Signed-off-by: Andy Whitcroft <apw@canonical.com>
    Signed-off-by: Tim Gardner <tim.gardner@canonical.com>

diff --git a/net/rds/ib_send.c b/net/rds/ib_send.c
index 84d90c9..4edcb1e 100644
--- a/net/rds/ib_send.c
+++ b/net/rds/ib_send.c
@@ -501,7 +501,7 @@ int rds_ib_xmit(struct rds_connection *conn, struct rds_message *rm,
 	int flow_controlled = 0;
 	int nr_sig = 0;
 
-	BUG_ON(off % RDS_FRAG_SIZE);
+	BUG_ON(!conn->c_loopback && off % RDS_FRAG_SIZE);
 	BUG_ON(hdr_off != 0 && hdr_off != sizeof(struct rds_header));
 
 	/* Do not send cong updates to IB loopback */
